sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc

```diff
@@ -0,0 +1,270 @@
+```
+
+There is no mechanism to decorate an OpenCL C kernel with a type-name, so the
+forms of `kernel_bundle::has_kernel` or `kernel_bundle::get_kernel` that take a
+type-name are not useful for kernels defined in OpenCL C.
+
+
+== Example
+
+The following example shows a simple SYCL program that defines an OpenCL C
+kernel as a string and then compiles and launches it.
+
+```
+#include <sycl/sycl.hpp>
+namespace syclex = sycl::ext::oneapi::experimental;
+
+int main() {
+  sycl::queue q;
+
+  // Kernel defined as an OpenCL C string.  This could be dynamically
+  // generated instead of a literal.
+  std::string source = R"""(
+    __kernel void my_kernel(__global int *in, __global int *out) {
+      size_t i = get_global_id(0);
+      out[i] = in[i]*2 + 100;
+    }
+  )""";
+
+  sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src =
+    syclex::create_kernel_bundle_from_source(
+      q.get_context(),
+      syclex::source_language::opencl,
+      source);
+
+  // Compile and link the kernel from the source definition.
+  sycl::kernel_bundle<sycl::bundle_state::executable> kb_exe =
+    syclex::build(kb_src);
+
+  // Get a "kernel" object representing the kernel defined in the
+  // source string.
+  sycl::kernel k = kb_exe.ext_oneapi_get_kernel("my_kernel");
+
+  constexpr int N = 4;
+  int input[N] = {0, 1, 2, 3};
+  int output[N] = {};
+
+  sycl::buffer inputbuf(input, sycl::range{N});
+  sycl::buffer outputbuf(output, sycl::range{N});
+
+  q.submit([&](sycl::handler &cgh) {
+    sycl::accessor in{inputbuf, cgh, sycl::read_only};
+    sycl::accessor out{outputbuf, cgh, sycl::read_write};
+
+    // Each argument to the kernel is a SYCL accessor.
+    cgh.set_args(in, out);
+
+    // Invoke the kernel over a range.
+    cgh.parallel_for(sycl::range{N}, k);
+  });
+}
+```
+
+
+== Issues
+
+* How should we expose the difference between OpenCL C versions?
+  It seems like there are two aspects to the problem.
+  Applications need some way to query which versions the backend (or device)
+  supports.
+  Applications also need some way to tell the runtime which version the kernel
+  is written in.
++
+--
+One option is to define separate enumerators in `source_language` for each
+version like this:
+
+```
+enum class source_language : /*unspecified*/ {
+  opencl_1_0,
+  opencl_1_1,
+  opencl_2_0,
+  opencl_3_0,
+};
+```
+
+Applications could then query the supported versions via
+`is_source_kernel_bundle_supported`, and applications would identify the
+version of their kernel string via the `lang` parameter to
+`create_kernel_bundle_from_source`.
+
+Alternatively, this extension could define just a single language enumerator
+(`opencl`), but also provide as separate query to get the supported OpenCL C
+versions.
+When building a kernel bundle, applications would be required to pass "-cl-std"
+via the `build_options` property in order to identify the OpenCL C version of
+their source string.
```

Comment:
It occurred to me that we will have the same issue with the CUDA language, which also has different versions.  It would be best if we chose a solution that works for both OpenCL and CUDA versions.

It seems like applications would want a way to query the supported versions.  Otherwise, it's trial-and-error.  Imagine an application that wants to use OpenCL 3.0 (or CUDA SM 90) if it's available, and fall back to something older if it's not available.  With no query, you'd have to keep trying to compile kernels until you found one that worked.

Reply:
The trial-and-error approach is what I had in mind 😆 .  I think most developers are going to hard-code some version, because they know what version of the language their code uses, and it doesn't matter if there are other versions (because they can't use them).  If they do have some other source code around, they're going to want to try it in some order, and I don't think trying to compile each of those versions until one succeeds is that bad.

Can we just defer to the native backend interface, here?  I don't think there's a general solution.  To check if a source string can actually be compiled, we need to check:

- **OpenCL**: CL_PLATFORM_VERSION (to get OpenCL version), CL_PLATFORM_EXTENSIONS (to get extensions) or CL_PLATFORM_EXTENSIONS_WITH_VERSION (to get extensions with versions, if OpenCL version is 3.x).  We'd probably need to check the device equivalents (CL_DEVICE_OPENCL_C_VERSION, CL_DEVICE_EXTENSIONS) as well.

- **CUDA**: I don't know the exact attributes to check, but I think you need to look at both the capabilities of the compiler (i.e., which version of the CUDA language can be compiled, currently 12.2) and the capabilities of the device you're targeting (i.e., whether the CUDA runtime can generate code for your SM version).

Reply:
@Pennycook, I'd also like to leave this issue unresolved in this PR and address it in a follow-on one.  My plan is to add a query API that tells the supported OpenCL versions, and we'll require the user to pass "-cl-std" as a build option, indicating the OpenCL version of their code.

As I noted in the comment above, @cperkinsintel has an implementation already for the spec in its current state, so I'd like the get the spec and implementation merged now.  We can add the new query APIs to the spec / implementation in the next sprint.

Reply:
Sounds good to me.